File: 322. Coin Change/322. Coin Change.py
class Solution(object):
  def coinChange(self, coins, amount):
    """
    :type coins: List[int]
    :type amount: int
    :rtype: int
    """
    if amount == 0:
      return 0
    
    coins.sort()
    
    dp = [-1 for i in range(amount+1)]
    
    for i in range(1, amount+1):
      if i in coins:
        dp[i] = 1
      else:
        for j in coins:
          if j > i:
            break
          else:
            if dp[i - j] != -1 and (dp[i] == -1 or dp[i] > dp[i-j] + 1):
              dp[i] = dp[i-j] + 1
    return dp[amount]

File: 322. Coin Change/322. Coin Change.rs
impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        let amount = amount as usize;
        const MAX: i32 = i32::max_value();
        let mut dp: Vec<i32> = vec![MAX; amount + 1];
        dp[0] = 0;
        for i in 1..=amount {
            for &coin in &coins {
                let coin = coin as usize;
                if coin <= i && dp[i - coin] < MAX {
                    dp[i] = dp[i].min(dp[i - coin] + 1);
                }
            }
        }
        return if dp[amount] < MAX { dp[amount] } else { -1 };
    }
}
